### PR TITLE
MM-9274- Sort Users in Channel by status

### PR DESCRIPTION
**Pull request by stephenkiers**
Repro:
1) On pre-release, go to the Contributors team, then Reception channel
2) Click the person icon in the channel header

Expected: Users who are online are listed first, followed by those who are away, offline and in do not disturb mode

Observed: Users who are offline are listed first

#### Ticket Link
https://mattermost.atlassian.net/browse/ICU-668

#### Related PRs
https://github.com/mattermost/mattermost-redux/pull/388
https://github.com/mattermost/mattermost-api-reference/pull/329\

#### Perf Test Results
- on server with 13014 users, 1 team, 12 channels, 
- on Channel with 6986 users (most of which did not have a Status entry, so I am not sure if that will slow down more, or not)

Original code ran in ~20ms (120 on fresh server)
New code seems to be about 10 times slower. Completing in ~200 (300ms, on fresh server)

### CODE REVIEW

**Comment by stephenkiers** [comment]
Had to change base to fix Github bug where it didn't recognize the latest commit... weird*

**Comment by stephenkiers** [comment]
bump (is this proper etiquette here?)

**Comment by jwilander** [comment]
I usually just DM the person the review is waiting on

File: store/sqlstore/user_store.go [hunk 1]
```diff
@@ -397,7 +397,22 @@ func (us SqlUserStore) GetProfilesInChannel(channelId string, offset int, limit
...
+			WHERE
+				ChannelMembers.ChannelId = :ChannelId
+			ORDER BY CASE Status
+				WHEN 'online' THEN 1
+				WHEN 'away' THEN 2
+				WHEN 'dnd' THEN 4
+				ELSE 3
+			END
+			LIMIT :Limit OFFSET :Offset
+		`
```
Comment (jwilander) [inline store/sqlstore/user_store.go]: The query looks fine but I don't think we want to change the ordering for absolutely everything using this function. For the majority of the cases I'm pretty sure we want the old behaviour, isn't it just the one instance on the client where we want to sort by status?

I think you need to add a "sort" query parameter to the `GET /channels/{id}/members` API endpoint that will default to the old query and then use this query if "?sort=status" is present

For example, [the GET /emojis endpoint](https://github.com/mattermost/mattermost-server/blob/master/api4/emoji.go#L85)

Comment (jwilander) [reply store/sqlstore/user_store.go]: You'll also need to update the API reference to match the new functionality around here https://github.com/mattermost/mattermost-api-reference/blob/master/v4/source/users.yaml#L64

**Comment by jwilander** [comment]
@esethna how is this a bug? This is definitely not old functionality that has been broken, it's always sorted the way it is on master.

**Comment by stephenkiers** [comment]
I need to go through this with someone that is a Go test expert. Who would be the best for that? I have little faith in the testing. and I am not sure the implementation is correct either.

**Comment by jwilander** [comment]
Ping me sometime tomorrow or later this week and I can go through the tests with you

**Comment by hmhealey** [comment]
@stephenkiers I've added some unit tests to the app and api4 packages

**Comment by stephenkiers** [comment]
@hmhealey looks like the tests failed

**Comment by hmhealey** [comment]
@stephenkiers I think it needs a rebase since the tests that failed were on the EE side

**Comment by stephenkiers** [comment]
Would I need to do anything else other that rebase? Still seems to be having the error...

**Comment by hmhealey** [comment]
Removing PM review since this isn't testable without the web app code

**Comment by jasonblais** [comment]
@hmhealey Do we just need to resolve the build failure and get JW review one more time?

**Comment by hmhealey** [comment]
@jasonblais The build failure should be fixed now. It was just an issue with master. We just need Joram's review now

**Pull request merged**